#  Web Scraping Project -  Flipkart Apple Mobiles data

It is the path toward get-together information from the Internet. In fact, even copy sticking the sections of your primary tune is a kind of web scratching! Regardless, the words “web scratching” by and large imply a connection that incorporates computerization. A couple of destinations could do without it when customized scrubbers collect their data, while others would not worry.

Expecting you are scratching a page deliberately for informative items, you are presumably not going to have any issues. Considering everything, it is a keen idea to do some assessment isolated and guarantee that you’re not ignoring any Terms of Service before you start a gigantic degree project. To get comfortable with the legal pieces of web scratching, take a gander at Legal Perspectives on Scraping Data from The Modern Web.

> ## We will be scraping Flipkart

![](https://i.imgur.com/QcDHnjr.png)

**Flipkart** is one such eCommerce company that promotes online sales in India. It is not only for accessories, but for a wide range of other daily use items as well, that Flipkart remains as the leading choice of online shoppers. Flipkart, the most prominent eCommerce website in India and is empowering tons of Indian businesses to venture into the competitive online shopping industry. This multi-vendor website serves as the platform for vendors to display their product to the website’s visitors. It was initiated in 2007 and recently achieved a similarweb ranking of 133 globally and 9th  amongst the top Indian websites. Flipkart is the first eCommerce website in India that has reached the $1 billion mark.

We will be scraping inside the **Mobile section** and specifically try to get all the information about **Apple Mobiles**.

Here is an outline of all the steps we will follow:

1. Download the webpage using "request library
2. Parse the HTML source code using beautiful soup
3. Extract items description, rating, prices, review and image Urls from 8 pages
4. Compile and extracted information into Python lists and dictionaries
5. Extract and combine data from multiple pages
6. Save the extracted information to a CSV file.


By the end of the project, we will created a CSV file in the similar format:<br>
- Each record is located on a separate line, delimited by a line break (CRLF).  
For example:<br>
aaa,bbb,ccc CRLF <br>
zzz,yyy,xxx CRLF


First of all we will **import jovian** and keep on executing **jovian.commit** to save our notebook.

In [ ]:
!pip install jovian --upgrade --quiet

In [ ]:
import jovian

In [ ]:
# Execute this to save new versions of the notebook
jovian.commit(project="project-web-scraping")

<IPython.core.display.Javascript object>

[jovian] Updating notebook "yashchamp96/project-web-scraping" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/yashchamp96/project-web-scraping


'https://jovian.ai/yashchamp96/project-web-scraping'

Use the "Run" button to execute the code.

#### Now it's time to import all the libraries we are going to use in our project.
- **import requests**- *We are going to use this library to download our desired webpage.*<br>
- **import bs4**- *It will allows us to pull data out of HTML documents.*<br>
- **import pandas**- *Pandas library is use to create the data frame of our requirements from the webpage.*

In [ ]:
import requests
import pandas as pd
from bs4 import BeautifulSoup

In [ ]:
url = 'https://www.flipkart.com/mobiles/pr?sid=tyy%2C4io&otracker=categorytree&p%5B%5D=facets.brand%255B%255D%3DAPPLE'

Let's download our url in a variable *response* using the *requests* library.

In [ ]:
response = requests.get(url)

- Now we will check the status code of our url which is saved in a variable.<br>
- If the *request* was successful it will provide the value in the *range(200,299)* in the output

In [ ]:
response.status_code

200

- We can check the text of our downloaded url from *request* method by **.text** function.<br>
- Here we are going to check the length of the all the contents in the url.

In [ ]:
contents = response.text
len(contents)

501795

- We will check the first *1000* characters of the webpage.

In [ ]:
contents [:1000]

'<!doctype html><html lang="en"><head><link href="https://rukminim1.flixcart.com" rel="preconnect"/><link rel="stylesheet" href="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/css/app_modules.chunk.94b5e7.css"/><link rel="stylesheet" href="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/css/app.chunk.dd97f3.css"/><meta http-equiv="Content-type" content="text/html; charset=utf-8"/><meta http-equiv="X-UA-Compatible" content="IE=Edge"/><meta property="fb:page_id" content="102988293558"/><meta property="fb:admins" content="658873552,624500995,100000233612389"/><meta name="robots" content="noodp"/><link rel="shortcut icon" href="https://static-assets-web.flixcart.com/www/promos/new/20150528-140547-favicon-retina.ico"/><link type="application/opensearchdescription+xml" rel="search" href="/osdd.xml?v=2"/><meta property="og:type" content="website"/><meta name="og_site_name" property="og:site_name" content="Flipkart.com"/><link rel="apple-touch-icon" sizes="57x57" href="/appl

Let's save all the source code of our desired webpage in an **.html** extension.

In [ ]:
with open('flipkart.html', 'w', encoding="utf-8") as file:
    file.write(contents)

In [ ]:
with open('flipkart.html', 'r',encoding="utf-8") as f:
    source = f.read()

- We can *read* and *write* our source code using the above *functions*.<br>
- Above we are using (encoding="utf-8") is used to *encode* all *1,112,064* valid character code points which might be available in our source code.

In [ ]:
source[:1000]

'<!doctype html><html lang="en"><head><link href="https://rukminim1.flixcart.com" rel="preconnect"/><link rel="stylesheet" href="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/css/app_modules.chunk.94b5e7.css"/><link rel="stylesheet" href="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/css/app.chunk.dd97f3.css"/><meta http-equiv="Content-type" content="text/html; charset=utf-8"/><meta http-equiv="X-UA-Compatible" content="IE=Edge"/><meta property="fb:page_id" content="102988293558"/><meta property="fb:admins" content="658873552,624500995,100000233612389"/><meta name="robots" content="noodp"/><link rel="shortcut icon" href="https://static-assets-web.flixcart.com/www/promos/new/20150528-140547-favicon-retina.ico"/><link type="application/opensearchdescription+xml" rel="search" href="/osdd.xml?v=2"/><meta property="og:type" content="website"/><meta name="og_site_name" property="og:site_name" content="Flipkart.com"/><link rel="apple-touch-icon" sizes="57x57" href="/appl

> ## Extracting information using Beautiful Soup

- To extract information from our HTML source code of the webpage we will be using [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/) library.

In [ ]:
doc = BeautifulSoup(source, 'html.parser')

- We can check the **type** of our extracted information.

In [ ]:
type(doc)

bs4.BeautifulSoup

- Now we will check the **title tag** of our webpage

In [ ]:
title = doc.title
title

<title>Mobile Price List | Compare Mobiles on Buy Online @ Flipkart</title>

- To get the text inside that tag we will use **.text**

In [ ]:
title.text

'Mobile Price List | Compare Mobiles on Buy Online @ Flipkart'

- Finding the **html tag** present in our webpage, we will use **.html** and by **len()** function we will see how many **html tags** are present.

In [ ]:
html_tag = doc.html
len(html_tag)

2

- **.a** extension is used to find the first **a tag** from our source code.

In [ ]:
first = doc.a
first

<a href="/"><img alt="Flipkart" class="_2xm1JU" src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/flipkart-plus_8d85f4.png" title="Flipkart" width="75"/></a>

As we can see above there are **href, img, src and title tag** are present inside **a tag**.

First **img** tag is

In [ ]:
doc.img

<img alt="Flipkart" class="_2xm1JU" src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/flipkart-plus_8d85f4.png" title="Flipkart" width="75"/>

First **p**  tag

In [ ]:
doc.p

<p>Flipkart Internet Private Limited, </p>

Now we are going use **find_all** for getting all the same kind of tags present in our source code.

In [ ]:
all_link_tags = doc.find_all('a')

In [ ]:
len(all_link_tags)

71

In [ ]:
all_link_tags[:3]

[<a href="/"><img alt="Flipkart" class="_2xm1JU" src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/flipkart-plus_8d85f4.png" title="Flipkart" width="75"/></a>,
 <a class="_21ljIi" href="/plus">Explore<!-- --> <span class="_2FVHGh">Plus</span><img src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/plus_aef861.png" width="10"/></a>,
 <a class="_1_3w1N" href="/account/login?ret=/mobiles/pr%3Fsid%3Dtyy%252C4io%26otracker%3Dcategorytree%26p%255B%255D%3Dfacets.brand%25255B%25255D%253DAPPLE">Login</a>]

There are near about **70** a tags present in our source code and we have seen **first 3** above.

> Same we will do for **img tag**

In [ ]:
image_tag = doc.find_all('img')

In [ ]:
len(image_tag)

81

In [ ]:
image_tag[:3]

[<img alt="Flipkart" class="_2xm1JU" src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/flipkart-plus_8d85f4.png" title="Flipkart" width="75"/>,
 <img src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/plus_aef861.png" width="10"/>,
 <img class="_3U-Vxu" height="21" src="//static-assets-web.flixcart.com/www/linchpin/fk-cp-zion/img/fa_62673a.png"/>]

### To find through class
![](https://i.imgur.com/5q8OAWs.png)
* We will first find through **class_='_4rR01T'** which will give us information about all the **mobile names**.

In [ ]:
matching_tags = doc.find_all(class_='_4rR01T')

In [ ]:
len(matching_tags)

24

In [ ]:
matching_tags[:3]

[<div class="_4rR01T">APPLE iPhone SE (Black, 64 GB)</div>,
 <div class="_4rR01T">APPLE iPhone SE (White, 64 GB)</div>,
 <div class="_4rR01T">APPLE iPhone 12 (Blue, 128 GB)</div>]

As we can see there are **24** items present in our desired webpage and we will get the **text** out of it to add all of them in a list.

In [ ]:
names = []

for tag in matching_tags:
    names.append({ 'Mobile Phone': tag.text.strip()}) #using .append to add every name in the list

names[:5]

[{'Mobile Phone': 'APPLE iPhone SE (Black, 64 GB)'},
 {'Mobile Phone': 'APPLE iPhone SE (White, 64 GB)'},
 {'Mobile Phone': 'APPLE iPhone 12 (Blue, 128 GB)'},
 {'Mobile Phone': 'APPLE iPhone SE (Red, 64 GB)'},
 {'Mobile Phone': 'APPLE iPhone 12 (Black, 64 GB)'}]

* Using **class_='_396cs4 _3exPp9'** to get all the information about **image link** and **mobile phone names**.
![](https://i.imgur.com/48U4zDd.png)

In [ ]:
image_tag = doc.find_all(class_='_396cs4 _3exPp9')

In [ ]:
len(image_tag)

24

In [ ]:
image_tag[:3]

[<img alt="APPLE iPhone SE (Black, 64 GB)" class="_396cs4 _3exPp9" src="https://rukminim1.flixcart.com/image/312/312/k9loccw0/mobile/p/z/q/apple-iphone-se-mxd02hn-a-original-imafrcpjfehbbqgb.jpeg?q=70"/>,
 <img alt="APPLE iPhone SE (White, 64 GB)" class="_396cs4 _3exPp9" src="https://rukminim1.flixcart.com/image/312/312/k9loccw0/mobile/6/b/z/apple-iphone-se-mxd12hn-a-original-imafrcqfsuzwa3dz.jpeg?q=70"/>,
 <img alt="APPLE iPhone 12 (Blue, 128 GB)" class="_396cs4 _3exPp9" src="https://rukminim1.flixcart.com/image/312/312/kg8avm80/mobile/y/7/n/apple-iphone-12-dummyapplefsn-original-imafwg8dqq7z8cgh.jpeg?q=70"/>]

In [ ]:
image_link_info = []
for link in image_tag:
    image_link_info.append({
        'Phone name': link['alt'],
        'url' : link['src']
    })
image_link_info[:3]

[{'Phone name': 'APPLE iPhone SE (Black, 64 GB)',
  'url': 'https://rukminim1.flixcart.com/image/312/312/k9loccw0/mobile/p/z/q/apple-iphone-se-mxd02hn-a-original-imafrcpjfehbbqgb.jpeg?q=70'},
 {'Phone name': 'APPLE iPhone SE (White, 64 GB)',
  'url': 'https://rukminim1.flixcart.com/image/312/312/k9loccw0/mobile/6/b/z/apple-iphone-se-mxd12hn-a-original-imafrcqfsuzwa3dz.jpeg?q=70'},
 {'Phone name': 'APPLE iPhone 12 (Blue, 128 GB)',
  'url': 'https://rukminim1.flixcart.com/image/312/312/kg8avm80/mobile/y/7/n/apple-iphone-12-dummyapplefsn-original-imafwg8dqq7z8cgh.jpeg?q=70'}]

* To get the information about description of all the phones we will use **class_='_1xgFaf'**

![](https://i.imgur.com/H8Rn0Wx.png)

In [ ]:
mobile_description = doc.find_all('ul', class_='_1xgFaf')
len(mobile_description)

24

In [ ]:
mobile_info_link=[]
for i in mobile_description:
    mobile_info_link.append({'Description': i.text.strip()})
mobile_info_link[:3]

[{'Description': '64 GB ROM11.94 cm (4.7 inch) Retina HD Display12MP Rear Camera | 7MP Front CameraA13 Bionic Chip with 3rd Gen Neural Engine ProcessorWater and Dust Resistant (1 meter for Upto 30 minutes, IP67)Fast Charge CapableWireless charging (Works with Qi Chargers | Qi Chargers are Sold SeparatelyBrand Warranty of 1 Year'},
 {'Description': '64 GB ROM11.94 cm (4.7 inch) Retina HD Display12MP Rear Camera | 7MP Front CameraA13 Bionic Chip with 3rd Gen Neural Engine ProcessorWater and Dust Resistant (1 meter for Upto 30 minutes, IP67)Fast Charge CapableWireless charging (Works with Qi Chargers | Qi Chargers are Sold SeparatelyBrand Warranty of 1 Year'},
 {'Description': '128 GB ROM15.49 cm (6.1 inch) Super Retina XDR Display12MP + 12MP | 12MP Front CameraA14 Bionic Chip with Next Generation Neural Engine ProcessorCeramic ShieldIndustry-leading IP68 Water ResistanceAll Screen OLED Display12MP TrueDepth Front Camera with Night Mode, 4K Dolby Vision HDR RecordingBrand Warranty for 1 Y

## Now we will make helper functions for all the information required.

>  #### 1- For image link information

In [ ]:
def get_image_link(doc):
    image_tag = doc.find_all(class_='_396cs4 _3exPp9')
    image_link_info = []
    for link in image_tag:
        image_link_info.append({'Phone name': link['alt'],
                            'url' : link['src']
                           })
    return image_link_info

> #### 2- For mobile phone names

In [ ]:
def mobile_name(doc):
    prase = doc.find_all(class_='_4rR01T')
    names = []
    for tag in prase:
        names.append({ 'Mobile Name': tag.text.strip()})
    return names

> #### 3- For mobile description

In [ ]:
def mobile_description(doc):
    mobile_description = doc.find_all('ul', class_='_1xgFaf')
    mobile_info_link=[]
    for i in mobile_description:
        mobile_info_link.append({'Description': i.text.strip()})
    return mobile_info_link

> #### 4- For prices of all the phones

In [ ]:
def prices(doc):
    price = doc.find_all(class_='_30jeq3 _1_WHN1')
    price_link = []
    for i in price:
        price_link.append({'Prices':i.text.strip()})
    return price_link

> #### 5- For the amount of discount available

In [ ]:
def percentage_off(doc):
    off = doc.find_all('div',class_='_3Ay6Sb')
    off_link =[]
    for i in off:
        off_link.append({'Percentage off':i.text.strip()})
    return off_link

> #### 6- For the overall rating of the mobile phones

In [ ]:
def rating(doc):
    ratings = doc.find_all('div', class_='_3LWZlK')
    all_ratings = []
    for i in ratings:
        all_ratings.append({'Ratings': i.text.strip()})
    return all_ratings

## Creating a helper function to add the information in .csv format

In [ ]:
def in_csv(items, path):
    # Open the file in write mode
    with open(path, 'w') as f:
        # Return if there's nothing to write
        if len(items) == 0:
            return

        # Write the headers in the first line
        headers = list(items[0].keys())
        f.write(','.join(headers) + '\n')

        # Write one item per line
        for item in items:
            values = []
            for header in headers:
                values.append(str(item.get(header, "")))
            f.write(','.join(values) + "\n")

### Creating csv files
We are creating different **csv** files for all the information using  our **in_csv** function.
- Mobile image links
- Mobile description
- Mobile names
- Mobile prices
- Mobile percentage off
- Mobile ratings


In [ ]:
link = get_image_link(doc)
in_csv(link,'image link.csv')

In [ ]:
desc = mobile_description(doc)
in_csv(desc,'description.csv')

In [ ]:
mobile_names = mobile_name(doc)
in_csv(mobile_names,'mobile name.csv')

In [ ]:
all_prices = prices(doc)
in_csv(all_prices,'prices.csv')

In [ ]:
all_percentage = percentage_off(doc)
in_csv(all_percentage,'Percentage off')

In [ ]:
all_ratings = rating(doc)
in_csv(all_ratings,'ratings.csv')

Now we can check all our csv file `File > Open `

Using the read function to see what's inside our **ratings.csv** file.

In [ ]:
with open('ratings.csv', 'r') as f:
    print(f.read())

Ratings
4.5
4.5
4.7
4.5
4.7
4.5
4.5
4.5
4.5
4.5
4.5
4.7
4.5
4.5
4.5
4.7
4.5
4.5
4.5
4.5
4.5
4.5
4.8
4.5



# Let's extract all the information using single function.

In [ ]:
base_url = 'https://www.flipkart.com/mobiles/pr?sid=tyy%2C4io&otracker=categorytree&p%5B%5D=facets.brand%255B%255D%3DAPPLE&page='

def mobile_name(doc_new):
    prase = doc_new.find_all('div', class_='_4rR01T')
    names = []
    for tag in prase:
        names.append(tag.text.strip())
    return names

def prices(doc_new):
    price = doc_new.find_all('div',class_='_30jeq3 _1_WHN1')
    price_link = []
    for i in price:
        price_link.append(i.text.strip())
    return price_link

def percentage_off(doc_new):
    off = doc_new.find_all('div',class_='_3Ay6Sb')
    off_link =[]
    for i in off:
        off_link.append({'Percentage off':i.text.strip()})
    return off_link

def rating(doc_new):
    ratings = doc_new.find_all('div', class_='_3LWZlK')
    all_ratings = []
    for i in ratings:
        all_ratings.append(i.text.strip())
    return all_ratings

def mobile_description(doc_new):
    mobile_description = doc_new.find_all('ul', class_='_1xgFaf')
    mobile_info_link=[]
    for i in mobile_description:
        mobile_info_link.append(i.text.strip())
    return mobile_info_link

def for_all_pages(page_no):
    page_no_url = base_url +  str(page_no)
    response = requests.get(page_no_url)

    if response.status_code != 200:
        print('Status code:', response.status_code)
        raise Exception('Failed to fetch web page ' + page_no_url)

    doc_new =  BeautifulSoup(response.text)
    mobile_names_new = mobile_name(doc_new)
    all_prices_new = prices(doc_new)
    all_ratings_new = rating(doc_new)
    first_page_new = mobile_description(doc_new)
    data_new = {'Name': mobile_names_new,  #getting all the information in a dictionary
       'Prices':all_prices_new,
       'ratings':all_ratings_new,
       'Description':first_page_new}

    return pd.DataFrame(data_new)   #return directly a dataframe.




As you can notice using only one function **for_all_pages** we can extract our information in **DataFrame** of all the pages we required.

In [ ]:
#Note- Run it twice if shows any error
for_all_pages(1) #information about first page.

Name     Prices ratings  \
0               APPLE iPhone SE (Black, 64 GB)    ₹30,699     4.5   
1               APPLE iPhone SE (White, 64 GB)    ₹30,199     4.5   
2               APPLE iPhone 12 (Blue, 128 GB)    ₹64,999     4.7   
3                 APPLE iPhone SE (Red, 64 GB)    ₹30,599     4.5   
4               APPLE iPhone 12 (Black, 64 GB)    ₹59,999     4.7   
5          APPLE iPhone 12 Mini (White, 64 GB)    ₹49,399     4.5   
6          APPLE iPhone 12 Mini (Black, 64 GB)    ₹52,999     4.5   
7           APPLE iPhone 12 Mini (Blue, 64 GB)    ₹50,099     4.5   
8          APPLE iPhone 12 Mini (Blue, 128 GB)    ₹54,999     4.5   
9              APPLE iPhone SE (Black, 128 GB)    ₹36,399     4.5   
10             APPLE iPhone SE (White, 128 GB)    ₹34,999     4.5   
11                APPLE iPhone 12 (Red, 64 GB)    ₹59,999     4.7   
12          APPLE iPhone 12 Mini (Red, 128 GB)    ₹54,399     4.5   
13               APPLE iPhone SE (Red, 128 GB)    ₹34,999     4.5   
14       APPLE iPhone 12 Mini (Purple, 128 GB)    ₹54,999     4.5   
15               APPLE iPhone 12 (Red, 128 GB)    ₹64,999     4.7   
16        APPLE iPhone 12 Mini (Purple, 64 GB)    ₹52,099     4.5   
17        APPLE iPhone 12 Mini (Black, 256 GB)    ₹64,299     4.5   
18        APPLE iPhone 12 Mini (Black, 128 GB)    ₹54,399     4.5   
19         APPLE iPhone 12 Mini (Blue, 256 GB)    ₹67,399     4.5   
20        APPLE iPhone 12 Mini (White, 256 GB)    ₹67,199     4.5   
21        APPLE iPhone 12 Mini (White, 128 GB)    ₹54,399     4.5   
22  APPLE iPhone 13 Pro Max (Graphite, 256 GB)  ₹1,39,900     4.8   
23               APPLE iPhone SE (Red, 256 GB)    ₹45,399     4.5   

                                          Description  
0   64 GB ROM11.94 cm (4.7 inch) Retina HD Display...  
1   64 GB ROM11.94 cm (4.7 inch) Retina HD Display...  
2   128 GB ROM15.49 cm (6.1 inch) Super Retina XDR...  
3   64 GB ROM11.94 cm (4.7 inch) Retina HD Display...  
4   64 GB ROM15.49 cm (6.1 inch) Super Retina XDR ...  
5   64 GB ROM13.72 cm (5.4 inch) Super Retina XDR ...  
6   64 GB ROM13.72 cm (5.4 inch) Super Retina XDR ...  
7   64 GB ROM13.72 cm (5.4 inch) Super Retina XDR ...  
8   128 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
9   128 GB ROM11.94 cm (4.7 inch) Retina HD Displa...  
10  128 GB ROM11.94 cm (4.7 inch) Retina HD Displa...  
11  64 GB ROM15.49 cm (6.1 inch) Super Retina XDR ...  
12  128 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
13  128 GB ROM11.94 cm (4.7 inch) Retina HD Displa...  
14  128 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
15  128 GB ROM15.49 cm (6.1 inch) Super Retina XDR...  
16  64 GB ROM13.72 cm (5.4 inch) Super Retina XDR ...  
17  256 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
18  128 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
19  256 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
20  256 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
21  128 GB ROM13.72 cm (5.4 inch) Super Retina XDR...  
22  256 GB ROM17.02 cm (6.7 inch) Super Retina XDR...  
23  256 GB ROM11.94 cm (4.7 inch) Retina HD Displa...

### Adding dataframes
Now we will create a helper function which will add all the information of **all the pages(num_of_page)** pages using **pd.concat()**.

In [ ]:
def all_info(num_of_page):
    # define an empty list
    all_df = []
    # Run a loop from 1 to n+1
    for i in range(1, num_of_page+1):
        all_df.append(for_all_pages(page_no=i))
        # Outside the loop variable you can directly merge the df and        return it in the same line.
    return pd.concat(all_df)

In [ ]:
final = all_info(9)

In [ ]:
final

Name     Prices ratings  \
0                      APPLE iPhone SE (Black, 64 GB)    ₹30,699     4.5   
1                      APPLE iPhone SE (White, 64 GB)    ₹30,199     4.5   
2                      APPLE iPhone 12 (Blue, 128 GB)    ₹64,999     4.7   
3                        APPLE iPhone SE (Red, 64 GB)    ₹30,599     4.5   
4                      APPLE iPhone 12 (Black, 64 GB)    ₹59,999     4.7   
..                                                ...        ...     ...   
19                 APPLE iPhone XS Max (Gold, 512 GB)  ₹1,44,900     4.6   
20               APPLE iPhone 6s Plus (Silver, 32 GB)    ₹34,900     4.4   
21  Apple iPhone 11 (Yellow, 128 GB) (Includes Ear...    ₹54,900     4.6   
22           APPLE iPhone 11 Pro Max (Silver, 256 GB)  ₹1,31,900     4.7   
23                APPLE iPhone 6s (Rose Gold, 128 GB)    ₹55,999     4.5   

                                          Description  
0   64 GB ROM11.94 cm (4.7 inch) Retina HD Display...  
1   64 GB ROM11.94 cm (4.7 inch) Retina HD Display...  
2   128 GB ROM15.49 cm (6.1 inch) Super Retina XDR...  
3   64 GB ROM11.94 cm (4.7 inch) Retina HD Display...  
4   64 GB ROM15.49 cm (6.1 inch) Super Retina XDR ...  
..                                                ...  
19  512 GB ROM16.51 cm (6.5 inch) Super Retina HD ...  
20  32 GB ROM13.97 cm (5.5 inch) Retina HD Display...  
21  128 GB ROM15.49 cm (6.1 inch) Liquid Retina HD...  
22  256 GB ROM16.51 cm (6.5 inch) Super Retina XDR...  
23  128 GB ROM11.94 cm (4.7 inch) Retina HD Displa...  

[216 rows x 4 columns]

**final** is our dataframe of *216 rows* and *4 columns* of the required information about **Apple phones** present in the *first 9 pages* at **flipkart**.

### Final csv file

> Finally we can

In [ ]:
final.to_csv('final info.csv',index=False)

# Conclusion

- We successfuly got a our final result and covered interesting features that will be helpful for others web scraping project.
- We covered `Python`,`requests` and `BeautifulSoup` libraries and there uses in web scraping.
- Now we have exctracted all the information in a csv file it will helpuful to see the information about Iphone which we might bought in future.

# Future work

- Will try to scrape different football websites to get a dataset on the basis of speed, total matches played, total trophies won and so on.

### References

Here are some link to learn more about used library:
- [request](https://docs.python-requests.org/projects/requests-html/en/latest/)
- [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [html source code](https://www.flipkart.com/mobiles/pr?sid=tyy%2C4io&otracker=categorytree&p%5B%5D=facets.brand%255B%255D%3DAPPLE)
- [pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html)


Here are some tutorial on web scraping:

- [Web scraping project from scratch](https://www.youtube.com/watch?v=RKsLLG-bzEY)
- [web scraping with yahoo finance](https://towardsdatascience.com/web-scraping-yahoo-finance-477fe3daa852)
- [web scraping at code with harry](https://www.youtube.com/watch?v=uufDGjTuq34)